File: app/src/main/java/com/falcon/restaurants/network/restaurant/FetchRestaurantsEndPoint.java
package com.falcon.restaurants.network.restaurant;

import com.falcon.restaurants.network.RetrofitInterface;
import java.util.List;
import javax.inject.Inject;
import io.reactivex.Single;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class FetchRestaurantsEndPoint {

    private static final String TAG = "FetchRestaurantEndPoint";

    RetrofitInterface retrofitInterface;

    @Inject
    public FetchRestaurantsEndPoint(RetrofitInterface RetrofitInterface) {
        this.retrofitInterface = RetrofitInterface;
    }

    public interface Listener {
        void onFetchSuccess(List<RestaurantNet> restaurantNets);
        void onFetchFailed(Throwable e);
    }

    public void fetch(String maxUpdatedAt, final Listener listener) {

        Single<List<RestaurantNet>> single = retrofitInterface.getRestaurants(maxUpdatedAt);
        SingleObserver<List<RestaurantNet>> singleObserver = new SingleObserver<List<RestaurantNet>>() {
            @Override
            public void onSubscribe(Disposable d) {
                System.out.println(TAG + "onSubscribe: ");
            }

            @Override
            public void onSuccess(List<RestaurantNet> restaurantNets) {
                System.out.println(TAG + "onSuccess: ");
                listener.onFetchSuccess(restaurantNets);
            }

            @Override
            public void onError(Throwable e) {
                System.out.println(TAG + "onError: " + e.getLocalizedMessage());
                listener.onFetchFailed(e);
            }
        };

        single.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(singleObserver);
    }
}

File: app/src/main/java/com/falcon/restaurants/network/restaurant/FetchRestaurantsEndPoint.kt
package com.falcon.restaurants.network.restaurant
import android.annotation.SuppressLint
import com.falcon.restaurants.network.RetrofitInterface
import com.falcon.restaurants.utils.Logger
import javax.inject.Inject
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

open class FetchRestaurantsEndPoint @Inject constructor(val retrofitInterface: RetrofitInterface){

    val TAG: String = "FetchRestaurantEndPoint"

    interface Listener {
        fun onFetchSuccess(restaurantNets: List<RestaurantNet>)
        fun onFetchFailed(e: Throwable)
    }

    @SuppressLint("CheckResult")
    open fun fetch(maxUpdatedAt: String, listener: Listener) {

        val single: Single<List<RestaurantNet>> = retrofitInterface.getRestaurants(maxUpdatedAt)

        single.subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(
                    { restaurantNets -> Logger.log(TAG , "onSuccess: ")
                        listener.onFetchSuccess(restaurantNets)},
                    { throwable -> Logger.log(TAG , "onError: " + throwable.getLocalizedMessage())
                        listener.onFetchFailed(throwable)}
                )
    }

}
